tests/integration/release/parallel_test_launcher.py

```diff
@@ -0,0 +1,218 @@
+    """
+    try:
+        result = subprocess.run(
+            f"squeue --job {job_id}", shell=True, capture_output=True, text=True
+        )
+        return (
+            job_id in result.stdout
+        )  # The job ID is listed in squeue if it's running/pending
+    except subprocess.CalledProcessError as e:
+        print(f"Error checking job status: {e}")
+        return False
+
+
+def wait_for_all_jobs(job_ids: Sequence[str | None], poll_interval: int = 1800) -> None:
+    """
+    Waits for all jobs to complete, checking their status every `poll_interval` seconds.
+
+    Args:
+        job_ids (list): List of Slurm job IDs to monitor.
+        poll_interval (int): Time interval (in seconds) between status checks. Default is 1800 seconds (30 minutes).
+    """
+    remaining_jobs = set(
+        x for x in job_ids if x is not None
+    )  # Track jobs that are still running
+    while remaining_jobs:
+        print(f"Checking status of {len(remaining_jobs)} remaining jobs...")
+
+        completed_jobs = []
+        for job_id in remaining_jobs:
+            if not is_job_running(job_id):
+                print(f"Job {job_id} has finished.")
+                completed_jobs.append(job_id)
+
+        # Remove completed jobs from the remaining jobs set
+        remaining_jobs -= set(completed_jobs)
+
+        if remaining_jobs:
+            print(
+                f"{len(remaining_jobs)} jobs have not finished. Checking again in {poll_interval // 60} minutes..."
+            )
+            time.sleep(poll_interval)  # Wait for the specified interval before checking again
+
+
+def parse_outputs(output_dir: str, job_ids: list[str] | None = None) -> None:
+    """Parses the output files to see how many jobs failed, didn't complete, and succeeded
+    in the specified directory, prints this to screen, and writes a summary CSV with this info.
+
+    Args:
+        output_dir (str): Directory where Slurm output/error files are stored.
+        job_ids (list[str] | None): List of Slurm job IDs to check. If provided, will verify that all expected outputs are present.
+    """
+    job_info = []
+
+    output_files = glob.glob(f"{output_dir}/*.out")
+    if job_ids:
+        if len(output_files) < len(job_ids):
+            print(
+                f"Warning: Expected {len(job_ids)} job outputs, but found only {len(output_files)} in {output_dir}."
+            )
+            print(
+                "This may indicate that some outputs are still being written. Rerun parse_outputs after all outputs have been written"
+            )
+            return
+        else:
+            print("All expected output files found.")
+
+    # iterate through all files/jobs
+    for file in output_files:
+        with open(file, "r") as f:
+            lines = f.readlines()
+            if not lines:
+                continue  # Skip empty files
+            last_line = lines[-1].strip()
+
+        if not (last_line.startswith("=") and last_line.endswith("=")):
```

Comment:
How do you know this for sure?

Reply:
I don't love it personally, but Rajan also explained it to me and it's probably fine. Just seems a bit brittle compared to return codes or something.